Support Theme Inheritance

waechtjn opened the pull request:
Currently, folder-color-switcher will not work with the Mint-X-Dark color theme. It will try to find colored folder icons in Mint-X-Dark-[COLOR], which does not exist. However, Mint-X-Dark does not define a custom folder icon. Therefore, it is safe to use the colored versions of the (normal) Mint-X theme, from which Mint-X-Dark inherits.

Instead of fixing this issue directly, I thought it might be useful to implemented support for icon theme inheritance (as defined by the `Inherits` line in a theme's `index.theme`) in general, as it could make the extensions work with other icon themes as well.

Here is a quick overview of my changes:
- Implement logging which can be controlled using an environment variable. (This is not part of the theme inheritance implementation (sorry!) but I could not debug without it.)
- Create a `Theme` class. It contains all logic for looking colored icon versions up and is independent of both Nemo and Caja. Reading of `index.theme` is done using `ConfigParser`.
- Use the new `Theme` class in both extensions. For this, I also created a new base class for them, which should make it easier to keep both versions (Nemo and Caja) in sync.

I have also included support for home folders (as in @mtwebster 's pull request).

clefebvre commented:
It looks good. Sorry it took me so long to test it.

A couple of things:

- There's an issue in Caja: 

Traceback (most recent call last):
  File "/usr/share/caja-python/extensions/caja-folder-color-switcher.py", line 279, in get_file_items
    supported_colors = self.theme.get_supported_colors(paths)
  File "/usr/share/caja-python/extensions/caja-folder-color-switcher.py", line 183, in get_supported_colors
    icon_path = self.find_folder_icon(color, directory)
  File "/usr/share/caja-python/extensions/caja-folder-color-switcher.py", line 167, in find_folder_icon
    colored_theme = relevant_ancestor.sibling(color)
AttributeError: 'NoneType' object has no attribute 'sibling'
Traceback (most recent call last):
  File "/usr/share/caja-python/extensions/caja-folder-color-switcher.py", line 279, in get_file_items
    supported_colors = self.theme.get_supported_colors(paths)
  File "/usr/share/caja-python/extensions/caja-folder-color-switcher.py", line 183, in get_supported_colors
    icon_path = self.find_folder_icon(color, directory)
  File "/usr/share/caja-python/extensions/caja-folder-color-switcher.py", line 167, in find_folder_icon
    colored_theme = relevant_ancestor.sibling(color)
AttributeError: 'NoneType' object has no attribute 'sibling'

The nemo extension works fine and adds support for Mint-X-Dark.

Minor, but both extensions produce that kind of warning:

/usr/lib/python2.7/dist-packages/gi/types.py:282: RuntimeWarning: Mixin class caja-folder-color-switcher.ChangeFolderColorBase is an old style class, please update this to derive from "object".

waechtjn commented:
Thanks for reviewing the pull request!

I have fixed the old-style class issue and rebased my changes against 1.0.7.

Additionally, I have included a defensive programming check against the exception from the Traceback you provided. However, this probably does not resolve the issue. The problem is that I cannot reproduce it (on Linux Mint 17.3 MATE). Could you provide the logging output when starting Caja with `LOG_FOLDER_COLOR_SWITCHER=10 caja`? It is also weird that you only get the exception with Caja since the `folder.svg` look-up logic is the same in both extensions (copy & paste).

clefebvre commented:
Hi,

This looks good. I'll probably change the warnings into debug... it's quite normal for some themes to fail and I don't want to get the output without debugging it.

monsta commented:
Hey guys. Just got an update to 1.1.0 in my LMDE 2, now there are some warnings from Caja in ```~/.xsession-errors```:

```
(caja:8992): Gtk-WARNING **: Refusing to add non-unique action 'ChangeFolderColorMenu::"P"' to action group 'DirExtensionsMenuGroup'

(caja:8992): Gtk-CRITICAL **: IA__gtk_action_get_name: assertion 'GTK_IS_ACTION (action)' failed

(caja:8992): Gtk-CRITICAL **: IA__gtk_action_get_name: assertion 'GTK_IS_ACTION (action)' failed

(caja:8992): Gtk-CRITICAL **: IA__gtk_ui_manager_add_ui: assertion 'name != NULL || type == GTK_UI_MANAGER_SEPARATOR' failed

(caja:8992): Gtk-CRITICAL **: IA__gtk_action_get_name: assertion 'GTK_IS_ACTION (action)' failed

(caja:8992): Gtk-CRITICAL **: IA__gtk_action_get_name: assertion 'GTK_IS_ACTION (action)' failed

(caja:8992): Gtk-CRITICAL **: IA__gtk_ui_manager_add_ui: assertion 'name != NULL || type == GTK_UI_MANAGER_SEPARATOR' failed

(caja:8992): Gtk-WARNING **: Refusing to add non-unique action 'ChangeFolderColorMenu::"B"' to action group 'DirExtensionsMenuGroup'

(caja:8992): Gtk-CRITICAL **: IA__gtk_action_get_name: assertion 'GTK_IS_ACTION (action)' failed

(caja:8992): Gtk-CRITICAL **: IA__gtk_action_get_name: assertion 'GTK_IS_ACTION (action)' failed

(caja:8992): Gtk-CRITICAL **: IA__gtk_ui_manager_add_ui: assertion 'name != NULL || type == GTK_UI_MANAGER_SEPARATOR' failed

(caja:8992): Gtk-CRITICAL **: IA__gtk_action_get_name: assertion 'GTK_IS_ACTION (action)' failed

(caja:8992): Gtk-CRITICAL **: IA__gtk_action_get_name: assertion 'GTK_IS_ACTION (action)' failed

(caja:8992): Gtk-CRITICAL **: IA__gtk_ui_manager_add_ui: assertion 'name != NULL || type == GTK_UI_MANAGER_SEPARATOR' failed

(caja:8992): Gtk-WARNING **: Refusing to add non-unique action 'ChangeFolderColorMenu::"G"' to action group 'DirExtensionsMenuGroup'

(caja:8992): Gtk-CRITICAL **: IA__gtk_action_get_name: assertion 'GTK_IS_ACTION (action)' failed

(caja:8992): Gtk-CRITICAL **: IA__gtk_action_get_name: assertion 'GTK_IS_ACTION (action)' failed

(caja:8992): Gtk-CRITICAL **: IA__gtk_ui_manager_add_ui: assertion 'name != NULL || type == GTK_UI_MANAGER_SEPARATOR' failed

(caja:8992): Gtk-CRITICAL **: IA__gtk_action_get_name: assertion 'GTK_IS_ACTION (action)' failed

(caja:8992): Gtk-CRITICAL **: IA__gtk_action_get_name: assertion 'GTK_IS_ACTION (action)' failed

(caja:8992): Gtk-CRITICAL **: IA__gtk_ui_manager_add_ui: assertion 'name != NULL || type == GTK_UI_MANAGER_SEPARATOR' failed
```

It's not major since folder color changing still works, but it really spams the log every time I open the right-click menu...

monsta commented:
Alright, fixed now in https://github.com/linuxmint/folder-color-switcher/commit/bb03475721efa994bdc4f4f6f45fbbdf6592fcd7.
